designs/build_for_layers.md

```diff
@@ -0,0 +1,83 @@
+What is the problem?
+--------------------
+AWS Lambda Layer is a ZIP archive that contains libraries, a custom runtime, or other dependencies. Layers allow us to keep our deployment package small, which makes development easier. To include libraries in a layer, they need to be placed in pre decided folders. For more information check [here](https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html#configuration-layers-path). Once files are in proper folder structure it can be zipped and uploaded to S3 where it’ll be ready to be used by Lambda Function.
+
+Current layer support of aws-sam-cli expects layer folder to be in final condition which then is zipped and uploaded to S3. That means if one need to compile java code, they have to do that outside of their sam-cli workflow. This design is solving this problem by including layer building process in sam build which will reduce the overhead to build layer separately outside of sam workflow.
+
+Out-of-Scope
+------------
+1. Implementation of Makefile based build. That is already in progress under [PR-166](https://github.com/awslabs/aws-lambda-builders/pull/166)
+
+User Experience Walkthrough
+---------------------------
+1. Add Metadata (check proposal for more details) to the layers in template.yaml file to opt-in to build for layers.
+1. Provide ManifestPath if applicable.
+1. Run `sam build` to build all functions and layers.
+1. Run `sam build FunctionLogicalID` to build single Function and all layers this function depend on.
+1. Run `sam build LayerLogicalID` to build single layer.
+
+Implementation
+==============
+### Proposal
+Build for layers will build Layer from aws-sam-cli and prepare artifacts in expected folder structure. Check [this](https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html#configuration-layers-path) for more details around expected folder structure.
+Build for layers will choose one of the following types of build workflows:
+#### NoBuildWorkflow
+This is not exactly a workflow. However, if user do not want to build their Layer using build for layers, then they don’t have to do anything. This feature is opt-in and if no `Metadata -> BuildMethod` is provided then we’ll keep following current workflow.
+
+#### StandardBuildWorkflow
+Sample YAML:
+```
+MyLayer:
+    Type: AWS::Serverless::LayerVersion
+    Properties:
+      ContentUri: my_layer
+      CompatibleRuntimes:
+        - python3.8
+    Metadata:
+      BuildMethod: python3.8 (or nodejs8.10 etc..)
```

Comment:
lets remove mention of nodejs8.10 in the example, as the compatbile runtimes are python3.8

Reply:
Theoretically it is possible to have this scenario. One can build layer using node, but say layer is generating jsom which then can be consumed by a lambda with python runtime.